### PR TITLE
Add support for Spark TimestampType using SchemaOptions.withSparkTimestamps

### LINKED ISSUE #111: ScalaPB is unable to convert timestamp column to protobuf timestamp
Doing so will result in
```
Can't extract value from timestamp#6: need struct type but got timestamp;
org.apache.spark.sql.AnalysisException: Can't extract value from timestamp#6: need struct type but got timestamp;
	at org.apache.spark.sql.catalyst.expressions.ExtractValue$.apply(complexTypeExtractors.scala:73)
	at org.apache.spark.sql.catalyst.analysis.Analyzer$$anonfun$resolveExpression$1.applyOrElse(Analyzer.scala:1154)
	at org.apache.spark.sql.catalyst.analysis.Analyzer$$anonfun$resolveExpression$1.applyOrElse(Analyzer.scala:1145)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformUp$4(TreeNode.scala:284)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:69)
```

I created a test repo here: https://github.com/Veske/protobuf-timestamp    
Run the ConvertTimestamp test and verify that it succeeds (`AnalysisException` is thrown)

### PR DESCRIPTION
**Pull request by mvitaly**
This should resolve #111 and #119

### CODE REVIEW

**Comment by mvitaly** [comment]
@thesamet I see that you closed #119 few days ago - this PR should resolve that feature request.
I would be happy to adjust the PR if you have comments

File: sparksql-scalapb/src/main/scala/org/apache/spark/sql/catalyst/expressions/timestampExpressions.scala [hunk 1]
```diff
@@ -0,0 +1,48 @@
+package org.apache.spark.sql.catalyst.expressions
```
Comment (thesamet) [inline sparksql-scalapb/src/main/scala/org/apache/spark/sql/catalyst/expressions/timestampExpressions.scala:1]: The two methods here look like they are just a special case of `Invoke` - are they really needed? I think they can be replaced with functions that return an instance of `Invoke`.

File: sparksql-scalapb/src/main/scala/scalapb/spark/FromCatalystHelpers.scala [hunk 2]
```diff
@@ -42,7 +26,17 @@ trait FromCatalystHelpers {
         cmp.scalaDescriptor.fields.map { fd =>
           fieldFromCatalyst(cmp, fd, input)
         }
-      else {
+      else if (schemaOptions.sparkTimestamps && cmp.scalaDescriptor.fullName == "google.protobuf.Timestamp") {
```
Comment (thesamet) [inline sparksql-scalapb/src/main/scala/scalapb/spark/FromCatalystHelpers.scala:29]: I'd prefer a more generic solution than testing for specific field types. My preference would be to have a map in schema options between `GeneratedMessageCompanion[_]` to functions that can convert that Spark datatype to protobuf and backwards. The code here can look into that map and call `Invoke` on these functions.

File: sparksql-scalapb/src/main/scala/scalapb/spark/TypedEncoders.scala [hunk 3]
```diff
@@ -41,11 +41,13 @@ trait TypedEncoders extends FromCatalystHelpers with ToCatalystHelpers with Seri
 
       val read = Invoke(reads, "read", ObjectType(classOf[Function[_, _]]))
 
-      Invoke(read, "apply", ObjectType(ct.runtimeClass), expr :: Nil)
+      val ret = Invoke(read, "apply", ObjectType(ct.runtimeClass), expr :: Nil)
+      ret
```
Comment (thesamet) [inline sparksql-scalapb/src/main/scala/scalapb/spark/TypedEncoders.scala:45]: Please keep code changes only to what required for the new feature.

**Comment by thesamet** [comment]
Thanks for putting together this PR! I'd be happy to work with you to get it in. What I'm looking for is a solution where users get to customize how protobuf types are converted to Spark (using datatype transforms, not catalyst expressions). Then, the timestamp conversion become a special case of that. I'd like the conversion logic to be centered in one place (right now, we have timestamp conversion in 3 places - fromCatalyst, toCatalyst, pMessageToRow)

**Comment by thesamet** [comment]
Hi @mvitaly - let me know if you'd like to continue and work on this, or we can archive for now.

**Comment by mvitaly** [comment]
> Hi @mvitaly - let me know if you'd like to continue and work on this, or we can archive for now.

I'm working on this as time permits. Made some progress with generic customization of types, but it's not in a working state yet.

**Pull request closed without merging**